File: clients/client-go/timers.go
package queen

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// The timer surface (PLAN_KV_TIMERS.md §4, §8.1): a message you schedule now and
// the broker pushes later, into a real queue, with the ordinary delivery
// guarantees from that moment on.
//
// FOUR THINGS THIS API CANNOT HIDE FROM YOU, and every one of them has cost
// somebody a production incident somewhere:
//
//  1. deliverAt is "NO EARLIER THAN", never "exactly at". The floor on this
//     stack is a single hop (p50 ~10 ms, fsync ~4 ms) plus one sweeper cycle.
//
//  2. `absent` on a cancel MAY MEAN ALREADY DELIVERED (§4.4). The row is
//     DELETEd on fire, so there is no tombstone: absent says "no longer
//     pending" and nothing more. The authority is the log -- look for the
//     timer's txn in the destination queue, which is why the cancel answer
//     carries that txn back. Any saga that cancels a compensation timer must
//     have the compensation consumer check the saga's KV state before acting.
//
//  3. `too_late` is a verdict, not a failure (§4.3). The broker holding the
//     claim has already packed that payload and is about to commit it; granting
//     the cancel would leave "did it go out?" with no answer. Bounded by the
//     lease (30 s by default). The remedy is a new key, or waiting for delivery
//     and acting on the message.
//
//  4. There is NO dedup net on the fire (§20.2, §20.7). Delete-plus-push in one
//     transaction is what makes a single fire exactly-once; a fixed txn is the
//     message's IDENTITY, not a net. Rescheduling or republishing a timer that
//     has already fired produces a SECOND message in the log, and no layer stops
//     it.

// Statuses a timer operation can answer. Closed taxonomy (§4.1): a client that
// has to tell them apart writes a switch, it does not read a sentence.
const (
	TimerStatusScheduled   = "scheduled"
	TimerStatusRescheduled = "rescheduled"
	TimerStatusCancelled   = "cancelled"
	TimerStatusAbsent      = "absent"
	TimerStatusTooLate     = "too_late"
)

// ---------------------------------------------------------------------------
// Requests.
// ---------------------------------------------------------------------------

// TimerSchedule describes one timer.
//
// Identity is (queue, timerKey) inside your tenant: scheduling the same key
// twice is an upsert, so a retry after a crash is safe by construction, and a
// rescheduled timer is a NEW timer under an old name (attempts and last error go
// back to zero).
type TimerSchedule struct {
	// Queue is the destination queue the message will be pushed into. Required.
	Queue string
	// TimerKey is the identity of the timer inside that queue. Required.
	TimerKey string
	// Partition is the destination partition ("Default" when empty).
	Partition string

	// Delay is how long from now the message becomes deliverable. A delay in the
	// PAST is legal and fires on the first cycle.
	//
	// Milliseconds on the wire, and that is a rule rather than an accident
	// (§20.6): the durations that can be sub-second are in milliseconds, the
	// ones that cannot are in seconds. A 250 ms retry backoff is one of the best
	// uses of a timer; a sub-second TTL is nobody's use case.
	Delay time.Duration

	// DeliverAt is client-side sugar for Delay: it is converted to a relative
	// delay when the request is sent, because an absolute instant is NOT
	// expressible on this wire -- one clock, Postgres's, so no broker skew can
	// enter (§4.2). Supplying both is an error.
	DeliverAt time.Time

	// Payload is the message body. []byte and json.RawMessage travel verbatim;
	// anything else is JSON-encoded first. It reaches the wire base64-encoded.
	Payload interface{}

	// TransactionID is the txn of the message that will be delivered, minted
	// when empty. It is what makes "was it sent?" answerable: look for it in the
	// destination queue. Every reschedule OVERWRITES it -- a rescheduled timer
	// is a new message (§20.2).
	TransactionID string
}

// TimerCancelOptions are the optional halves of a cancel.
type TimerCancelOptions struct {
	// ExpectedTransactionID is echoed back on `absent`, so the "has it already
	// been delivered?" check needs no second API. It is your own identifier and
	// nothing is revealed by sending it.
	ExpectedTransactionID string
}

// TimerListOptions are the optional halves of a list.
type TimerListOptions struct {
	// After is an EXCLUSIVE keyset cursor over timerKey, not an offset.
	After string
	// Limit defaults to 100 server-side and is CLAMPED, never rejected.
	Limit int
}

// TimerOp is one element of a POST /api/v1/timers batch or of a transaction's
// `timers` rider array. Build it with ScheduleTimerOp, RescheduleTimerOp or
// CancelTimerOp.
//
// The fields the SERVER owns are absent by construction and must stay absent:
// producerSub, messageId, tenant and deliverAt are all rejected rather than
// ignored. Ignoring producerSub would make a timer the one way in this product
// to forge the provenance of a frame (§4.2).
type TimerOp struct {
	Op        string `json:"op"`
	Queue     string `json:"queue"`
	TimerKey  string `json:"timerKey"`
	Partition string `json:"partition,omitempty"`
	DelayMs   *int64 `json:"delayMs,omitempty"`
	Txn       string `json:"txn,omitempty"`
	Payload   string `json:"payload,omitempty"`

	err error
}

// encodeTimerPayload turns a caller's payload into the base64 the wire carries.
func encodeTimerPayload(v interface{}) (string, error) {
	var raw []byte
	switch p := v.(type) {
	case nil:
		raw = []byte("null")
	case []byte:
		raw = p
	case json.RawMessage:
		raw = p
	case string:
		// A string payload is JSON-encoded, not taken as raw bytes: a consumer
		// reads messages as JSON, and passing "hello" through verbatim would
		// deliver a frame that is not valid JSON.
		b, err := json.Marshal(p)
		if err != nil {
			return "", err
		}
		raw = b
	default:
		b, err := json.Marshal(v)
		if err != nil {
			return "", fmt.Errorf("queen: timer payload cannot be encoded as JSON: %w", err)
		}
		raw = b
	}
	return base64.StdEncoding.EncodeToString(raw), nil
}

// scheduleOp builds a schedule or reschedule op out of a TimerSchedule.
func scheduleOp(kind string, s TimerSchedule) TimerOp {
	op := TimerOp{Op: kind, Queue: s.Queue, TimerKey: s.TimerKey, Partition: s.Partition}
	if s.Queue == "" {
		op.err = errors.New("queen: a timer needs a queue")
		return op
	}
	if s.TimerKey == "" {
		op.err = errors.New("queen: a timer needs a timerKey")
		return op
	}
	if s.Delay != 0 && !s.DeliverAt.IsZero() {
		op.err = errors.New("queen: set Delay or DeliverAt, not both")
		return op
	}
	delay := s.Delay
	if !s.DeliverAt.IsZero() {
		delay = time.Until(s.DeliverAt)
	}
	op.DelayMs = Int64(int64(delay / time.Millisecond))

	payload, err := encodeTimerPayload(s.Payload)
	if err != nil {
		op.err = err
		return op
	}
	op.Payload = payload

	op.Txn = s.TransactionID
	if op.Txn == "" {
		op.Txn = GenerateUUID()
	}
	return op
}

// ScheduleTimerOp builds a schedule operation.
func ScheduleTimerOp(s TimerSchedule) TimerOp { return scheduleOp("schedule", s) }

// RescheduleTimerOp builds a reschedule operation. It is the SAME upsert as a
// schedule -- the two names exist because they mean different things to the
// reader, not to the server.
func RescheduleTimerOp(s TimerSchedule) TimerOp { return scheduleOp("reschedule", s) }

// CancelTimerOp builds a cancel operation for a transaction's rider array (the
// saga close bundle: ack the message and cancel its compensation timer in one
// commit).
//
// For a STANDALONE cancel use Timers.Cancel, which goes to
// DELETE /api/v1/timers/:queue/*timerKey -- the one route that is guaranteed
// never to be blocked, whatever the cluster's quota state (§9.6). A cancel sent
// through POST /api/v1/timers inherits that route's class instead, and a mixed
// batch is refused WHOLE on a blocked cluster.
func CancelTimerOp(queue, timerKey string) TimerOp {
	op := TimerOp{Op: "cancel", Queue: queue, TimerKey: timerKey}
	if queue == "" || timerKey == "" {
		op.err = errors.New("queen: a cancel needs a queue and a timerKey")
	}
	return op
}

func validateTimerOps(ops []TimerOp) error {
	for i, op := range ops {
		if op.err != nil {
			return fmt.Errorf("queen: timer operation at index %d: %w", i, op.err)
		}
	}
	return nil
}

// ---------------------------------------------------------------------------
// Results.
// ---------------------------------------------------------------------------

// TimerResult is the answer to a schedule, reschedule or cancel.
//
// OK is false on `absent` and on `too_late`, and neither is a Go error: they are
// verdicts with HTTP 200. `absent` carrying OK:false is deliberate -- the
// in-house lesson is queue delete, where deleted:false with a 200 read as
// success to every client that trusted the field.
type TimerResult struct {
	OK       bool   `json:"ok"`
	Status   string `json:"status"`
	Queue    string `json:"queue"`
	TimerKey string `json:"timerKey"`
	// TransactionID is the txn of the message this timer will deliver (or was
	// going to deliver). On `absent` it is the txn to look for in the
	// destination queue.
	TransactionID string `json:"txn"`
	// MessageID is promised at schedule time so a delivered frame can be
	// correlated without a second API.
	MessageID string    `json:"messageId"`
	DeliverAt time.Time `json:"deliverAt"`

	// OpIndex and Type are filled when the result comes back inside a
	// transaction's flat result array.
	OpIndex int    `json:"opIndex"`
	Type    string `json:"type"`
}

// TimerInfo is one pending timer, as returned by Peek (with the payload) or List
// (without it).
type TimerInfo struct {
	Found         bool
	Queue         string
	TimerKey      string
	Partition     string
	DeliverAt     time.Time
	TransactionID string
	MessageID     string
	// Payload is the stored payload, base64-decoded, and EXACTLY as it is
	// stored: peek is an inspection surface and does not quietly decrypt or
	// decompress what the fire will deliver -- Encrypted and PayloadZstd tell
	// you the truth about the bytes. Nil on List, which never carries payloads.
	Payload     []byte
	PayloadZstd bool
	Encrypted   bool
	ProducerSub string
	// Attempts counts only PERMANENT and configuration failures: a transient one
	// (a serialization failure, a connection loss) backs off without consuming
	// the budget. At QUEEN_SWEEPER_MAX_ATTEMPTS the timer goes to the
	// destination queue's DLQ.
	Attempts  int
	LastError string
	// Claimed means a broker holds this row right now. A timer in BACKOFF reads
	// claimed:false, deliberately: it is still cancellable, and that is the
	// whole reason the claim token is cleared on failure.
	Claimed   bool
	CreatedAt time.Time
	UpdatedAt time.Time
}

// TimerPage is one page of a list.
type TimerPage struct {
	Rows      []TimerInfo
	Truncated bool
	NextAfter string
}

// timerInfoWire is the shape on the wire. It is separate from TimerInfo for one
// reason: `payload` arrives base64 and TimerInfo hands back bytes, and a public
// struct with a staging field would be a field that is always empty by the time
// anybody reads it.
type timerInfoWire struct {
	Found         bool      `json:"found"`
	Queue         string    `json:"queue"`
	TimerKey      string    `json:"timerKey"`
	Partition     string    `json:"partition"`
	DeliverAt     time.Time `json:"deliverAt"`
	TransactionID string    `json:"txn"`
	MessageID     string    `json:"messageId"`
	Payload       string    `json:"payload"`
	PayloadZstd   bool      `json:"payloadZstd"`
	Encrypted     bool      `json:"encrypted"`
	ProducerSub   string    `json:"producerSub"`
	Attempts      int       `json:"attempts"`
	LastError     string    `json:"lastError"`
	Claimed       bool      `json:"claimed"`
	CreatedAt     time.Time `json:"createdAt"`
	UpdatedAt     time.Time `json:"updatedAt"`
}

// timerPageWire is the list envelope. `rows` never carries a `found` field, so a
// row is always a real timer.
type timerPageWire struct {
	Rows      []timerInfoWire `json:"rows"`
	Truncated bool            `json:"truncated"`
	NextAfter string          `json:"nextAfter"`
}

func (w timerInfoWire) info(found bool) (TimerInfo, error) {
	info := TimerInfo{
		Found:         found,
		Queue:         w.Queue,
		TimerKey:      w.TimerKey,
		Partition:     w.Partition,
		DeliverAt:     w.DeliverAt,
		TransactionID: w.TransactionID,
		MessageID:     w.MessageID,
		PayloadZstd:   w.PayloadZstd,
		Encrypted:     w.Encrypted,
		ProducerSub:   w.ProducerSub,
		Attempts:      w.Attempts,
		LastError:     w.LastError,
		Claimed:       w.Claimed,
		CreatedAt:     w.CreatedAt,
		UpdatedAt:     w.UpdatedAt,
	}
	if w.Payload != "" {
		raw, err := base64.StdEncoding.DecodeString(w.Payload)
		if err != nil {
			return info, fmt.Errorf("queen: timer payload is not valid base64: %w", err)
		}
		info.Payload = raw
	}
	return info, nil
}

// ---------------------------------------------------------------------------
// The client surface.
// ---------------------------------------------------------------------------

// Timers is the timer API. Get it from Queen.Timers.
type Timers struct {
	httpClient *HttpClient
}

// Timers returns the timer API.
func (q *Queen) Timers() *Timers {
	return &Timers{httpClient: q.httpClient}
}

// timerPath builds a path route. See kvPath for why the segments are escaped and
// why a `/` inside a timerKey survives it.
func timerPath(queue, timerKey string) string {
	return "/api/v1/timers/" + url.PathEscape(queue) + "/" + strings.ReplaceAll(url.PathEscape(timerKey), "%2F", "/")
}

// timerBatchRequest is the batch envelope, the same shape as the KV one.
type timerBatchRequest struct {
	Operations []TimerOp `json:"operations"`
}

func (t *Timers) apply(ctx context.Context, ops []TimerOp) ([]TimerResult, error) {
	if err := validateTimerOps(ops); err != nil {
		return nil, err
	}
	if len(ops) == 0 {
		return nil, errors.New("queen: a timer batch needs at least one operation")
	}
	body, err := t.httpClient.PostRaw(ctx, "/api/v1/timers", timerBatchRequest{Operations: ops})
	if err != nil {
		return nil, surfaceError(err)
	}
	var out struct {
		Results []TimerResult `json:"results"`
	}
	if err := decodeKV(body, &out); err != nil {
		return nil, fmt.Errorf("queen: timers response: %w", err)
	}
	if len(out.Results) != len(ops) {
		return nil, fmt.Errorf("queen: timers returned %d results for %d operations", len(out.Results), len(ops))
	}
	return out.Results, nil
}

// Schedule schedules one timer.
func (t *Timers) Schedule(ctx context.Context, s TimerSchedule) (TimerResult, error) {
	res, err := t.apply(ctx, []TimerOp{ScheduleTimerOp(s)})
	if err != nil {
		return TimerResult{}, err
	}
	return res[0], nil
}

// Reschedule moves an existing timer. Identical to Schedule on the server: the
// same upsert, with attempts and last error reset.
func (t *Timers) Reschedule(ctx context.Context, s TimerSchedule) (TimerResult, error) {
	res, err := t.apply(ctx, []TimerOp{RescheduleTimerOp(s)})
	if err != nil {
		return TimerResult{}, err
	}
	return res[0], nil
}

// ScheduleBatch schedules several timers in one transaction. A (queue,
// timerKey) may appear at most once per call.
func (t *Timers) ScheduleBatch(ctx context.Context, schedules ...TimerSchedule) ([]TimerResult, error) {
	ops := make([]TimerOp, len(schedules))
	for i, s := range schedules {
		ops[i] = ScheduleTimerOp(s)
	}
	return t.apply(ctx, ops)
}

// Apply runs a mixed batch of timer operations in one transaction.
//
// Note the asymmetry this buys, and it is deliberate: on a cluster blocked for
// quota, a batch containing a schedule is refused WHOLE, cancels included. When
// what you need is the cancel, use Cancel (§9.6).
func (t *Timers) Apply(ctx context.Context, ops ...TimerOp) ([]TimerResult, error) {
	return t.apply(ctx, ops)
}

// Cancel removes a pending timer, through the route that is never blocked.
//
// Read the `absent` contract at the top of this file before trusting a false
// return: it does NOT mean the timer never fired.
func (t *Timers) Cancel(ctx context.Context, queue, timerKey string, opts ...TimerCancelOptions) (TimerResult, error) {
	if queue == "" || timerKey == "" {
		return TimerResult{}, errors.New("queen: a cancel needs a queue and a timerKey")
	}
	if len(opts) > 1 {
		return TimerResult{}, errors.New("queen: pass at most one TimerCancelOptions")
	}
	path := timerPath(queue, timerKey)
	if len(opts) == 1 && opts[0].ExpectedTransactionID != "" {
		path += "?txn=" + url.QueryEscape(opts[0].ExpectedTransactionID)
	}
	body, err := t.httpClient.DeleteRaw(ctx, path, nil)
	if err != nil {
		return TimerResult{}, surfaceError(err)
	}
	var r TimerResult
	if err := decodeKV(body, &r); err != nil {
		return TimerResult{}, fmt.Errorf("queen: timer cancel response: %w", err)
	}
	return r, nil
}

// Peek reads one pending timer, payload included. A miss is Found:false with no
// error.
func (t *Timers) Peek(ctx context.Context, queue, timerKey string) (TimerInfo, error) {
	body, err := t.httpClient.GetRaw(ctx, timerPath(queue, timerKey))
	if err != nil {
		return TimerInfo{}, surfaceError(err)
	}
	var w timerInfoWire
	if err := decodeKV(body, &w); err != nil {
		return TimerInfo{}, fmt.Errorf("queen: timer peek response: %w", err)
	}
	return w.info(w.Found)
}

// List pages through the pending timers of ONE queue.
//
// The queue is mandatory and is a path segment rather than a filter: a
// tenant-wide list would be a scan that an end user of your own product could
// trigger (§4.1).
func (t *Timers) List(ctx context.Context, queue string, opts ...TimerListOptions) (TimerPage, error) {
	if queue == "" {
		return TimerPage{}, errors.New("queen: a timer list needs a queue")
	}
	if len(opts) > 1 {
		return TimerPage{}, errors.New("queen: pass at most one TimerListOptions")
	}
	path := "/api/v1/timers/" + url.PathEscape(queue)
	if len(opts) == 1 {
		q := url.Values{}
		if opts[0].After != "" {
			q.Set("after", opts[0].After)
		}
		if opts[0].Limit > 0 {
			q.Set("limit", strconv.Itoa(opts[0].Limit))
		}
		if len(q) > 0 {
			path += "?" + q.Encode()
		}
	}
	body, err := t.httpClient.GetRaw(ctx, path)
	if err != nil {
		return TimerPage{}, surfaceError(err)
	}
	var w timerPageWire
	if err := decodeKV(body, &w); err != nil {
		return TimerPage{}, fmt.Errorf("queen: timer list response: %w", err)
	}
	page := TimerPage{Truncated: w.Truncated, NextAfter: w.NextAfter}
	for _, row := range w.Rows {
		info, err := row.info(true)
		if err != nil {
			return TimerPage{}, err
		}
		page.Rows = append(page.Rows, info)
	}
	return page, nil
}

File: clients/client-py/queen/timers/timers.py
"""
The timer API -- ``client.timers`` and ``client.timer(queue)``.

PLAN_KV_TIMERS.md §4 (contract), §8.1 (routes), §9.6 (why cancel has a route of
its own, which makes it an SDK decision and not only a server one).

THE THREE SENTENCES THAT HAVE TO TRAVEL WITH THIS API:

  1. ``deliverAt`` IS "NOT BEFORE", NEVER "EXACTLY AT". The measured floor on
     this stack is a single hop around 10 ms plus one sweep cycle. A healthy
     timer lands within about ten milliseconds of the sweeper's minimum sleep;
     past its maximum sleep (1 s) there is a wake-up problem, not a load
     problem.
  2. ``absent`` MEANS "NO LONGER PENDING" AND MAY MEAN ALREADY DELIVERED
     (§4.4). There is no tombstone -- a delivered timer has no row. The
     authority is the log: look for the timer's ``txn`` in the destination
     queue, which is why every answer carries it back.
  3. ``too_late`` IS A VERDICT, NOT A FAILURE (§4.3). The broker holding the
     claim has already unpacked and packed that payload and is about to commit
     it. The remedy is a new key, or waiting for delivery and acting on the
     message.
"""

from __future__ import annotations

from typing import Any, Dict, Iterable, List, Optional
from urllib.parse import quote, urlencode

import httpx

from ..errors import TimerError, wrap_http_error
from ..utils import logger
from . import ops as _ops

_PATH = "/api/v1/timers"


class TimerResult(dict):
    """A timer result element, which is a plain dict plus a truthful ``bool``.

    ``status`` is a CLOSED taxonomy -- ``scheduled | rescheduled | cancelled |
    absent | too_late`` -- closed because a client that has to tell them apart
    writes a match, not a substring search. ``__bool__`` reads ``ok``, so the
    three verdicts that are not successes (``absent``, ``too_late``, and a
    ``cancel`` that found nothing) do not read as one.
    """

    __slots__ = ()

    def __bool__(self) -> bool:
        if "ok" in self:
            return bool(self["ok"])
        if "found" in self:
            return bool(self["found"])
        return len(self) > 0


class Timers:
    """Scheduled messages against one Queen deployment."""

    op = _ops

    def __init__(self, http_client: Any) -> None:
        self._http_client = http_client

    # -----------------------------------------------------------------
    # POST /api/v1/timers -- schedule and reschedule, batch.
    # -----------------------------------------------------------------

    async def batch(self, operations: Iterable[Dict[str, Any]]) -> List[TimerResult]:
        """Apply timer ops in one transaction. Results are INDEX-ALIGNED (§6.4).

        A cancel IS accepted in this array -- it is the same stored procedure --
        but a cancel sent here inherits this ROUTE's authorization class, so on
        a cluster that is over quota a mixed batch is refused whole (§9.6). Use
        :meth:`cancel`, which takes the route that is never blocked.
        """
        op_list = list(operations)
        if not op_list:
            return []
        for op in op_list:
            _ops.check_not_server_owned(op)
        logger.log("Timers.batch", {"count": len(op_list), "ops": [o.get("op") for o in op_list]})
        try:
            response = await self._http_client.post(_PATH, {"operations": op_list})
        except Exception as error:  # noqa: BLE001 - re-raised, possibly re-typed
            raise wrap_http_error(error, TimerError) from None
        results = (response or {}).get("results") or []
        return [TimerResult(r) if isinstance(r, dict) else r for r in results]

    async def _one(self, op: Dict[str, Any]) -> TimerResult:
        results = await self.batch([op])
        if not results:
            raise TimerError(
                "the server returned no result for a single-op call",
                request=httpx.Request("POST", _PATH),
                response=httpx.Response(500),
                code="timers_result_missing",
            )
        return results[0]

    async def schedule(
        self,
        queue: str,
        timer_key: str,
        payload: Any,
        *,
        delay_ms: Optional[Any] = None,
        delay: Optional[Any] = None,
        partition: Optional[str] = None,
        txn: Optional[str] = None,
        payload_zstd: bool = False,
    ) -> TimerResult:
        """Schedule a message for later. Upsert on ``(queue, timer_key)``.

        Returns ``{ok, status, queue, timerKey, txn, messageId, deliverAt}``.
        The ``messageId`` is promised at schedule time so the delivered frame
        can be correlated without a second API call.
        """
        return await self._one(
            _ops.schedule(
                queue, timer_key, payload,
                delay_ms=delay_ms, delay=delay, partition=partition, txn=txn, payload_zstd=payload_zstd,
            )
        )

    async def reschedule(
        self,
        queue: str,
        timer_key: str,
        payload: Any,
        *,
        delay_ms: Optional[Any] = None,
        delay: Optional[Any] = None,
        partition: Optional[str] = None,
        txn: Optional[str] = None,
        payload_zstd: bool = False,
    ) -> TimerResult:
        """The same upsert under the name of the intent. ``attempts`` returns to
        zero: a rescheduled timer is a new timer under an old name."""
        return await self._one(
            _ops.schedule(
                queue, timer_key, payload,
                delay_ms=delay_ms, delay=delay, partition=partition, txn=txn,
                payload_zstd=payload_zstd, reschedule=True,
            )
        )

    # -----------------------------------------------------------------
    # DELETE /api/v1/timers/:queue/*timerKey -- §9.6, never blockable.
    # -----------------------------------------------------------------

    async def cancel(self, queue: str, timer_key: str, *, txn: Optional[str] = None) -> TimerResult:
        """Cancel a pending timer, over the route that cannot be blocked.

        THIS IS NOT A DETAIL OF THE URL. ``POST /api/v1/timers`` carries cancels
        in the same array as schedules, so a cancel sent there inherits the
        schedule's authorization and is refused on a cluster that is over quota
        or has scheduling paused. The fire never switches itself off, so a
        tenant that cannot cancel keeps producing messages it cannot stop until
        the horizon or an operator -- the block would produce the exact opposite
        of its purpose. Hence the separate route, and hence this method.

        ``txn`` is echoed back on ``absent`` so the "was it already delivered?"
        check needs no second API (§4.4). Pass the one you got from
        :meth:`schedule`.

        Cancelling a timer that is in BACKOFF succeeds: during backoff the claim
        token is NULL on purpose.
        """
        path = f"{_PATH}/{quote(queue, safe='')}/{quote(timer_key, safe='/')}"
        if txn:
            path = f"{path}?{urlencode({'txn': txn})}"
        logger.log("Timers.cancel", {"queue": queue, "timer_key": timer_key})
        try:
            response = await self._http_client.delete(path)
        except Exception as error:  # noqa: BLE001
            raise wrap_http_error(error, TimerError) from None
        return TimerResult(response or {})

    # -----------------------------------------------------------------
    # Reads.
    # -----------------------------------------------------------------

    async def peek(self, queue: str, timer_key: str) -> TimerResult:
        """One key, with its payload exactly as stored.

        A miss is ``{"found": false}`` with HTTP 200. ``encrypted`` tells the
        truth about the payload -- peek is an inspection surface and does not
        quietly decrypt what the fire will deliver as an envelope.
        """
        path = f"{_PATH}/{quote(queue, safe='')}/{quote(timer_key, safe='/')}"
        try:
            response = await self._http_client.get(path)
        except Exception as error:  # noqa: BLE001
            raise wrap_http_error(error, TimerError) from None
        return TimerResult(response or {})

    async def list(
        self,
        queue: str,
        *,
        after: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Keyset page of the pending timers of ONE queue.

        The queue is mandatory and is a path segment rather than a filter on
        purpose (§4.1): a tenant-wide list would be a scan that an end user of
        the customer could trigger, on the first endpoint of this product whose
        call rate is decided by somebody else's web traffic.

        ``after`` is an EXCLUSIVE keyset cursor, not an offset. ``limit`` is
        clamped by the server and never rejected, with ``truncated`` telling
        the truth.
        """
        params: Dict[str, Any] = {}
        if after:
            params["after"] = after
        if limit is not None:
            params["limit"] = limit
        path = f"{_PATH}/{quote(queue, safe='')}"
        if params:
            path = f"{path}?{urlencode(params)}"
        try:
            return await self._http_client.get(path) or {}
        except Exception as error:  # noqa: BLE001
            raise wrap_http_error(error, TimerError) from None

    async def list_all(self, queue: str, *, limit: Optional[int] = None, max_rows: Optional[int] = None) -> List[Dict[str, Any]]:
        """Walk ``nextAfter`` to the end of a queue's pending timers."""
        rows: List[Dict[str, Any]] = []
        after: Optional[str] = None
        while True:
            page = await self.list(queue, after=after, limit=limit)
            rows.extend(page.get("rows") or [])
            if max_rows is not None and len(rows) >= max_rows:
                return rows[:max_rows]
            if not page.get("truncated"):
                return rows
            after = page.get("nextAfter")
            if not after:
                return rows

    # -----------------------------------------------------------------
    # The builder.
    # -----------------------------------------------------------------

    def timer(self, queue: str) -> "TimerBuilder":
        """Fluent form, with ``schedule()`` / ``reschedule()`` / ``cancel()``
        as the terminals."""
        return TimerBuilder(self, queue)


class TimerBuilder:
    """Fluent timer, mirroring ``QueueBuilder``'s shape in this client.

        await (client.timer("orders")
                     .key("order:9f1:expire")
                     .payload({"orderId": "9f1"})
                     .after(timedelta(minutes=30))
                     .schedule())

    The terminals are coroutines here and are NOT coroutines on the transaction
    variant (``tx.timer(...)``), which returns the transaction for chaining --
    the same split this client already has between ``client.queue(q).push()``
    and ``tx.queue(q).push()``.
    """

    def __init__(self, timers: Timers, queue: str) -> None:
        self._timers = timers
        self._queue = queue
        self._key: Optional[str] = None
        self._payload: Any = None
        self._has_payload = False
        self._delay_ms: Optional[Any] = None
        self._delay: Optional[Any] = None
        self._partition: Optional[str] = None
        self._txn: Optional[str] = None
        self._payload_zstd = False

    def key(self, timer_key: str) -> "TimerBuilder":
        self._key = timer_key
        return self

    def payload(self, payload: Any, *, zstd: bool = False) -> "TimerBuilder":
        self._payload = payload
        self._has_payload = True
        self._payload_zstd = zstd
        return self

    def after_ms(self, delay_ms: Any) -> "TimerBuilder":
        self._delay_ms = delay_ms
        return self

    def after(self, delay: Any) -> "TimerBuilder":
        """A ``timedelta``. There is no string duration parser in this SDK and
        there will not be one: Python already has the type."""
        self._delay = delay
        return self

    def partition(self, partition: str) -> "TimerBuilder":
        self._partition = partition
        return self

    def txn(self, txn: str) -> "TimerBuilder":
        self._txn = txn
        return self

    def _require_key(self) -> str:
        if not self._key:
            raise ValueError("timer needs .key(<timerKey>)")
        return self._key

    def _require_payload(self) -> Any:
        if not self._has_payload:
            raise ValueError("timer needs .payload(<data>)")
        return self._payload

    async def schedule(self) -> TimerResult:
        return await self._timers.schedule(
            self._queue, self._require_key(), self._require_payload(),
            delay_ms=self._delay_ms, delay=self._delay, partition=self._partition,
            txn=self._txn, payload_zstd=self._payload_zstd,
        )

    async def reschedule(self) -> TimerResult:
        return await self._timers.reschedule(
            self._queue, self._require_key(), self._require_payload(),
            delay_ms=self._delay_ms, delay=self._delay, partition=self._partition,
            txn=self._txn, payload_zstd=self._payload_zstd,
        )

    async def cancel(self) -> TimerResult:
        """Terminal cancel -- takes the DELETE route, the one that is never
        blockable (§9.6)."""
        return await self._timers.cancel(self._queue, self._require_key(), txn=self._txn)
